src/client.ts

```diff
@@ -284,148 +147,246 @@ export class Client {
+          name: axiosError.name,
+          message: axiosError.message
+        };
       }
 
-      if (connectOpts.keepTryingOnFailure && this.autoReconnect) {
-        if (retriable) {
-          await wait(connectOpts.retryDelay);
-          return this.connect(connectOpts);
-        }
+      this.logger.error('Failed to connect streaming client', { error });
+      throw err;
+    }
+  }
+
+  private backoffConnectRetryHandler (connectOpts: StreamingClientConnectOptions, err: any, connectionAttempt: number): boolean {
+    // if we exceed the `numOfAttempts` in the backoff config it still calls this retry fn and just ignores the result
+    // if that's the case, we just want to bail out and ignore all the extra logging here.
+    if (!connectOpts.keepTryingOnFailure) {
+      return false;
+    }
+
+    const additionalErrorDetails: any = { connectionAttempt, error: err };
+
+    if (err.name === 'AxiosError') {
+      const axiosError = err as AxiosError;
+      const config = axiosError.config;
+      let sanitizedError = {
+        config: {
+          url: config.url,
+          method: config.method
+        },
+        status: axiosError.response?.status,
+        code: axiosError.code,
+        name: axiosError.name,
+        message: axiosError.message
+      };
 
-        this.logger.error('Streaming client received and error that it can\'t recover from and will not attempt to reconnect', { error: e });
+      additionalErrorDetails.error = sanitizedError;
+
+      if ([401, 403].includes(err.response?.status || 0)) {
+        this.logger.error('Streaming client received an error that it can\'t recover from and will not attempt to reconnect', additionalErrorDetails);
+        return false;
       }
+    }
+
+    // if we get a sasl error, that means we made it all the way to the point of trying to open a websocket and
+    // it was rejected for some reason. At this point we should do a hard reconnect then try again.
+    if (err instanceof SaslError) {
+      this.logger.info('hardReconnectRequired set to true due to sasl error');
+      this.hardReconnectRequired = true;
+      Object.assign(additionalErrorDetails, { channelId: err.channelId, stanzaInstanceId: err.stanzaInstanceId });
+    }
+
+    // we don't need to log the stack for a timeout message
+    if (err instanceof TimeoutError) {
+      additionalErrorDetails.error = err.message;
 
-      throw e;
+      const details = (err as any).details;
+      if (details) {
+        additionalErrorDetails.details = details;
+      }
     }
+
+    this.logger.error('Failed streaming client connection attempt, retrying', additionalErrorDetails, { skipServer: err instanceof OfflineError });
+    return true;
   }
 
-  _tryToConnect () {
-    this.startServerLogging();
-    this.connecting = true;
+  private async makeConnectionAttempt () {
+    if (!navigator.onLine) {
+      throw new OfflineError('Browser if offline, skipping connection attempt');
```

Comment:
Typo -> "is"

Reply:
I already asked this in a different comment, but leaving another here. Is there a specific reason you didn't just use `Error` here?